Задание №3

Istomininik opened the pull request:
Создан класс Triad и созданы методы по заданию

Istomininik committed:
Задание №3

Istomininik committed:
Маленькое изменение

Изменён файл Triad.cpp

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	third = 0;
+}
+
+Triad::Triad(int fr, int sc, int th)
```
если входящие параметры не изменяются в функции, то они должны быть `const`

MarinaPaley replied:
и я второй раз пишу, добавьте `const`

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+
+string Triad::toString()
+{
+    cout << "\n Тройка чисел: " << first << "; " << second << "; " << third << endl;
```
не в консоль!
используйте 
```cpp
#include <sstream>
// ...

std::stringstream ss;

//put arbitrary formatted data into the stream
ss << 4.5 << ", " << 4 << " whatever";

//convert the stream buffer into a string
std::string str = ss.str();

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	third = th;
+}
+
+string Triad::toString()
```
по ссылке!

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+    return string();
+}
+
+int Triad::operator1(Triad& Triad)
```
в помойку метод

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	third = 1900;
+}
+
+Date::Date(int fr, int sc, int th)
```
необходимо проверять на валидность введеных чисел, используя сеттеры

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+
+void Date::setDay(int fr)
+{
+	first = fr;
```
зачем сеттер, если он ничего не проверяет?

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+
+void Date::setYear(int th)
+{
+	third = th;
```
зачем сеттер, если он ничего не проверяет?

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+
+void Date::setMonth(int sc)
+{
+	second = sc;
```
зачем сеттер, если он ничего не проверяет?

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	third = th;
+}
+
+int Date::getDay()
```
если метод класса не изменяет содержимое объекта, то он должен быть помчен как `const`

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	return first;
+}
+
+int Date::getMonth()
```
если метод класса не изменяет содержимое объекта, то он должен быть помчен как `const`

MarinaPaley replied:
и где исправления?

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+	return second;
+}
+
+int Date::getYear()
```
если метод класса не изменяет содержимое объекта, то он должен быть помчен как `const`

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+
+
+
+bool Date::checkDay(Date& date) {
```
так вот это и нужно проверять в сеттерах, ОТДЕЛЬНО!

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+        (checkMonth(date) == true) &&
+        (checkYear(date) == true)) 
+    {
+        cout << "This date is correct!\n";
```
мы кидаем исключение! а не вывод в консоль!

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -0,0 +1,199 @@
+}
+
+
+int Date::operator_n(Date& date)
```
я не буду это смотреть, это не понятно ... ни по смыслу того, что должно быть, ни, тем более по реализации

MarinaPaley commented on Task.3/Triad.h:
```diff
@@ -0,0 +1,58 @@
+﻿#pragma once
+#include <string>
+
+class Triad
```
где документация?

MarinaPaley commented on Task.3/Triad.h:
```diff
@@ -0,0 +1,58 @@
+	int getYear();
+
+
+	bool checkDay(Date& date);
+
+	bool checkMonth(Date& date);
+
+	bool checkYear(Date& date);
+
+	void checkDate(Date& date);
+
```
убрать в сеттеры

MarinaPaley commented on Task.3/Triad.h:
```diff
@@ -0,0 +1,58 @@
+	void checkDate(Date& date);
+
+
+	int operator_n(Date& date);
```
вообще не понимаю, что это такое и зачем нужно

Istomininik committed:
Переработка всего проекта.

Переделан метод увеличения даты на n дней.
Добавлено три метода для увеличения каждого числа на 1.
Косметические изменения.

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
-    return string();
+    stringstream ss;
+    ss << first << ". " << second << ". " << third;
+    string str = ss.str();
```
`return ss.str();`

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
+    third = th;
 }
 
+int days_in_month[12] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
```
почему `int`?

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
+
 
-void Date::setDay(int fr)
+void Date::setDay(int fr, bool a, Date& date) throw(std::invalid_argument)
```
неправильная сигнатура метода
ООП не понимаете

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
 {
-	first = fr;
+    first = fr;
+    if (a)
```
нейминг должен быть говорящим!. что такое `a? я не понимаю ни сигнатуры, ни реализации!

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
 }
 
-void Date::setMonth(int sc)
+void Date::setMonth(int sc, Date& date) throw(std::invalid_argument)
```
неправильная сигнатура

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
+void Date::setMonth(int sc, Date& date) throw(std::invalid_argument)
 {
-	second = sc;
+    if (sc >= 1 && sc <= 12)
```
константы в коде - зло!

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
+void Date::setYear(int th) throw(std::invalid_argument)
 {
-	third = th;
+    if (th >= 0)
```
сначала проверка на неправильность и выброс исключения, потом присваивание значения полю

MarinaPaley commented on Task.3/Triad.cpp:
```diff
@@ -1,199 +1,161 @@
 }
 
-void Date::setYear(int th)
+void Date::setYear(int th) throw(std::invalid_argument)
```
`const`

MarinaPaley reviewed (changes requested):
абсолютно неправильная логика создания объекта Дата
в конструкторе необходимо вызывать правильные сеттеры (у Вас неправильная сигнатура и реализация)
пишите самостоятельно, предыдущий чей-то `pull requset` был также плох

Istomininik committed:
Поправки в программе

Изменены сеттеры
